File: Programming Languages/JavaScript/Core/syntax.js

// ==============================
// ========== COMMENTS ==========
// ==============================

// single line comment starting with //

/* multi-line comment
can span multiple lines
*/



// ===============================
// ========== VARIABLES ==========
// ===============================

// Delacaration
let block_variable; // block-scoped variable, only accessible within the block they are defined in like function, if statement, loop, etc.
var function_variable; // function-scoped variable, accessible within the function it is defined even if defined within a block
const constant_variable = null; // block-scoped constant, must be initialized at declaration and cannot be reassigned
let inline_declaration = 42; // inline declaration and initialization for both let and var
let var1, var2 = 'Hello', var3="key3"; // multiple declarations in a single statement
let [a=0, b, ...rest] = [1, 2, 3]; // array destructuring assignment, assigns every element of the array to corresponding variable, with rest operator collecting remaining unassigned elements into an array which can be empty, additional variables without corresponding elements will be assigned undefined or default values if provided
let {key1=1, key2:b, [var3], ...rest} = {key1: 1, key2: 2, key3: 3}; // object destructuring assignment, assigns values of object properties to corresponding variables (which can be renamed using :) by matching property names (which can be evaluted if not direct value using []), with rest operator collecting remaining unassigned properties into an object which can be empty, additional variables without corresponding properties will be assigned undefined or default values if provided

// Types and Literals
let int = 10_000_00; // integer number, underscores can be used as separators for better readability
let float = 10.5; // floating-point number, underscores can be used like integers
let infinity = Infinity; // special numeric value representing infinity
let nan = NaN; // special numeric value representing "Not-a-Number"
let bigint = 9007199254740991n; // BigInt for representing integers larger than 2^53 - 1 using 'n' suffix or using BigInt() constructor, underscores can be used like integers
let string_double = "Hello, World!"; // string using double quotes
let string_single = 'Hello, World!'; // string using single quotes
let template_string = `Hello, ${string_single}!`; // template string using backticks allows for interpolation of values using ${}
let boolean = true; // boolean value, can be either true or false
let null_value = null; // special value representing "no value"
let undefined_value = undefined; // special value representing "not assigned"
let array = [1, 2, 3]; // array, an ordered collection of values
let object = { key: 'value', "string key": 42 }; // object, a collection of key-value pairs, keys can only be strings or valid identifiers which convert to strings implicitly



// ===============================
// ========== OPERATORS ==========
// ===============================

// Arithmetic Operators
let sum = 5 + 3; // addition
let difference = 5 - 3; // subtraction
let product = 5 * 3; // multiplication
let quotient = 5 / 3; // division
let remainder = 5 % 3; // modulus
let exponentiation = 5 ** 3; // exponentiation
let prefix_increment = ++sum; // prefix increment, increments value before using it
let postfix_increment = sum++; // postfix increment, increments value after using it
let prefix_decrement = --difference; // prefix decrement, decrements value before using it
let postfix_decrement = difference--; // postfix decrement, decrements value after using it
let unary_minus = -product; // unary minus, negates the value
let unary_plus = +quotient; // unary plus, converts the operand to a number
let coalescing = null_value ?? 'default'; // nullish coalescing operator, returns right operand if left is null or undefined
let ternary = (5 > 3) ? 'yes' : 'no'; // ternary operator, returns value after ? if condition is true, otherwise returns value after :

// Comparison Operators
let equal = 5 == '5'; // loose equality, compares values after type coercion
let strict_equal = 5 === 5; // strict equality, compares values and types without type coercion
let not_equal = 5 != '5'; // loose inequality, compares values after type coercion
let strict_not_equal = 5 !== 5; // strict inequality, compares values and types without type coercion
let greater_than = 5 > 3; // greater than
let less_than = 5 < 3; // less than
let greater_than_equal = 5 >= 3; // greater than or equal to
let less_than_equal = 5 <= 3; // less than or equal to

// Logical Operators
let logical_and = true && false; // logical AND, returns false if first operand is false (short-circuit) or both are false, otherwise returns true
let logical_or = true || false; // logical OR, returns true if first operand is true (short-circuit), otherwise returns false if both are false 
let logical_not = !true; // logical NOT, negates the boolean value

// Bitwise Operators
let bitwise_and = 5 & 3; // bitwise AND
let bitwise_or = 5 | 3; // bitwise OR
let bitwise_xor = 5 ^ 3; // bitwise XOR
let bitwise_not = ~5; // bitwise NOT
let left_shift = 5 << 1; // left shift
let right_shift = 5 >> 1; // right shift
let unsigned_right_shift = 5 >>> 1; // unsigned right shift

// Assignment Operators
let assign = 10; // simple assignment
assign += 5; // addition assignment
assign -= 5; // subtraction assignment
assign *= 5; // multiplication assignment
assign /= 5; // division assignment
assign %= 5; // modulus assignment
assign **= 2; // exponentiation assignment
assign <<= 1; // left shift assignment
assign >>= 1; // right shift assignment
assign >>>= 1; // unsigned right shift assignment
assign &&= 'default'; // logical AND assignment, assigns the result of right operand && left operand to left operand
assign ||= 'default'; // logical OR assignment, assigns the result of right operand || left operand to left operand
assign &= 3; // bitwise AND assignment
assign |= 3; // bitwise OR assignment
assign ^= 3; // bitwise XOR assignment
assign ??= 'default'; // nullish coalescing assignment, assigns right operand if left is null or undefined

// Array Spread Operator
let array2 = [0, ...array, 4, 5]; // array spread operator, expands elements of an array into another array
function_with_params(...array); // using spread operator to pass array elements as individual arguments to a function



// ========================================
// ========== CONTROL STATEMENTS ==========
// ========================================

// If-Else Statement
if (var1 == 0) { // if statement, executes block if condition is true, {} are optional for single statement blocks
    console.log('Condition is zero');
} else if (var1 > 0) { // else if statement, just an if statement as a part of else block, will execute block if condition is true and all previous conditions are false, {} are optional for single statement blocks
    console.log('Condition is positive');
} else { // else statement, executes block if all previous conditions are false, {} are optional for single statement blocks
    console.log('Condition is negative');
}

// Switch Statement
switch (var2) { // switch statement, evaluates expression and executes corresponding case block
    case 'Hello': // case clause, compares switch expression to value using strict equality
        console.log('Greeting detected'); // block executed if case matches
        break; // break statement, exits the switch statement
    case 'See you': // case clause with empty block, falls through to next case without a break
    case 'Goodbye':
        console.log('Farewell detected');
        break;
    default: // default clause, executes if no case matches, optional
        console.log('No match found');
}


// Try-Catch-Finally Statement

// Throw Statement



// ===========================
// ========== LOOPS ==========
// ===========================

// For Loop
for (let i = 0; i < 5; i++) { // for loop, executes block n timees equal to (end - start) / step, first part is initialization, second part is condition, third part is increment/decrement, separated by semicolons, all parts are optional and can be empty, {} are optional for single statement blocks   
    console.log(i); // block executed in each iteration
    if (i === 2) 
        break; // break statement, exits the loop
    else if (i === 3)
        continue; // continue statement, skips the rest of the current iteration and moves to the next iteration
}

// Range-based For Loop
for (const element of [1, 2, 3]) { // range-based for loop, iterates over elements of iterable objects like arrays, strings, maps, sets, etc., each element is assigned to the loop variable in each iteration, {} are optional for single statement blocks, break and continue work like regular for loop
    console.log(element); // block executed in each iteration
}

// While Loop
while (true) { // while loop, executes block as long as condition is true, {} are optional for single statement blocks, break and continue work like regular for loop
    console.log('Infinite loop');
}

// Do-While Loop
do { // do-while loop, executes block at least once and then continues to execute as long as condition is true, {} are optional for single statement blocks, break and continue work like regular for loop
    console.log('This will run at least once');
} while (true); // condition is checked after the block is executed 



// ===============================
// ========== FUNCTIONS ==========
// ===============================

function function_declaration() { // function declaration with no parameters
    return 'Hello'; // return statement, returns value from function and exits the function
}
function function_with_params(param1, param2 = 'default', ...restParams) {
    function_with_params(...param1); // forwarding Rest parameters using spread syntax to another function 
} // function declaration with positional parameter, default parameter, and Rest parameters which collects remaining arguments into an array
const function_expression = function(){}; // anonymous function declation
const named_function_expression = function namedFunc() { // named function expression declaration 
    namedFunc(); // recursive call to the named function expression
}; 
const arrow_function = (param1, param2 = 'default', ...restParams) => 'Hello'; // arrow function with all types of parameters (can be by empty) and implicit return when body is an expression
const arrow_function_block = param => { console.log(param); }; // arrow function with a parameter (parentheses are optional if only one parameter) and a block body which can contain multiple statements and optional return



// =============================
// ========== CLASSES ==========
// =============================

// Class Declaration
class BaseClass { // class declaration
    static static_attribute = 'static value'; // static class attribute, shared among all instances and accessed using the class name, initialization is optional
    static #private_static_attribute = 'private static value'; // private static class attribute, only accessible within the class using # syntax, initialization is optional
    static { // static initialization block, executed once when the class is defined, can be used to define static attributes and methods or perform any setup tasks, can access static attributes and methods using class name or this keyword
        console.log('Static block executed'); // code executed when class is defined
        this.static_attribute2 = true; // defining another static attribute using this keyword
    }
    static static_method() { // static class method, shared among all instances and accessed using the class name, cannot access this keyword, parameters optional
        return this.static_attribute2 ? // access static attribute using class name
            BaseClass.static_attribute : null; // access static attribute using this keyword which refers to the class itself in static methods
    }
    static #private_static_method() { // private static class method, only accessible within the class using # syntax, parameters optional
        return this.#private_static_attribute; // access private static attribute using this keyword which refers to the class itself in static methods
    }

    #private_attribute = 4; // private instance attribute, only accessible within the class using # syntax, each instance has its own copy, initialization is optional
    _attribute = 3; // public instance attribute, accessible from outside the class, each instance has its own copy, initialization is optional
    constructor(param) { // constructor method, called when creating an instance with new keyword, parameters optional
        this.attribute2 = this._attribute; // creating instance attribute directly inside the constructor using this keyword which refers to the instance, initialization is mandatory
        this.attribute2 += this.#private_attribute; // access private instance attribute using this keyword which refers to the instance
    }
    get attribute() { // getter method for attribute, allows accessing attribute like a property while executing code, no parameters allowed, can be static for static attributes
        return this._attribute + 1; 
    }
    set attribute(value) { // setter method for attribute, allows assigning value to attribute like a property while executing code, must have exactly one parameter, can be static for static attributes
        this._attribute = value - 1;
    }
    #private_method() { // private instance method, only accessible within the class using # syntax, parameters optional
        return this.#private_attribute; // access private instance attribute using this keyword which refers to the instance
    }
    method(param) { // method declaration, can be called on instances of the class, parameters optional
        this.attribute = param; // assign to instance attribute to modify it using this keyword
        return this.attribute; // return statement optional, access instance attribute using this keyword
    }
    method2() { 
        this.method(this.attribute + 1); // calling another method of the same class using this keyword 
    }
}
let instance = new BaseClass(10); // creating an instance of the class using new keyword and passing parameters to constructor
instance?. // optional chaining, ?. checks if instance is not null or undefined before accessing method or attribute, otherwise returns undefined
method?.(20);  // optional chaining, ?. checks if method exists before calling it with parameters, otherwise returns undefined

// Inheritance
class DerivedClass extends BaseClass { // class declaration with inheritance using extends keyword
    constructor(param) { // constructor method in derived class, must call super() before using this keyword
        super(param); // call to parent class constructor using super() with all necessary parameters
        this.attribute2 = this.attribute + 1; // create new derived class access and assigne parent class attribute using this to it
    } 
    method(param) { // method overriding, replace parent class method with same name
        this.method2(); // call parent class method using this keyword (when not overridden)
        super.method(param); // call parent class method using super keyword 
    }
}



// =============================
// ========== MODULES ==========
// =============================

// Exporting

// Importing

File: Programming Languages/GO/Core/syntax.go

// ==============================
// ========== PACKAGES ==========
// ==============================

// Main Package
package main // used in the main file to indicate that this is the entry point of the program
package folder // used in every other file which must be the same name as the folder the file is in



// ===================================
// ========== MAIN FUNCTION ==========
// ===================================

func main() { // the main function is the entry point of the program, { must be in the same line as the function declaration

} // } can go anywhere



// ===============================
// ========== VARIABLES ==========
// ===============================

// Declaration
var variable int = 10  // variable declaration with a type and initialization, type is optional and can be infered if the variable is initialized, initialization is optional
const constant int = 10 // constant declaration with a type and initialization, type is optional and can be infered since intialization is mandatory
variable := 10 // short variable declaration, can only be used inside functions, type is inferred
var variable1, variable2 = 10, "Hello" // multiple variable declaration with initialization, types are inferred for each variable
var variable1, variable2 int = 10, 20 // multiple variable declaration with initialization and one type at the end, type is optional if all variables are initialized

// Types
var variable int // signed integer variable, is at least 32 bits in size, depends of the system, default initialized to 0
var variable int8 // 8-bit signed integer variable, range: [-128, 128[, default initialized to 0
var variable int16 // 16-bit signed integer variable, range: [-2^15, 2^15[, default initialized to 0
var variable int32 // 32-bit signed integer variable, range: [-2^31, 2^31[, default initialized to 0
var variable int64 // 64-bit signed integer variable, range: [-2^63, 2^63[, default initialized to 0
var variable uint // unsigned integer variable, is at least 32 bits in size, depends of the system, default initialized to 0
var variable uint8 // 8-bit unsigned integer variable, range: [0, 256[, default initialized to 0
var variable uint16 // 16-bit unsigned integer variable, range: [0, 2^16[, default initialized to 0
var variable uint32 // 32-bit unsigned integer variable, range: [0, 2^32[, default initialized to 0
var variable uint64 // 64-bit unsigned integer variable, range: [0, 2^64[, default initialized to 0
var variable float32 // 32-bit floating-point variable, default initialized to 0
var variable float64 // 64-bit floating-point variable, default initialized to 0
var variable bool // boolean variable, can only be true or false, default initialized to false
var variable string // string variable, default initialized to ""
var variable complex64 // 64-bit complex number variable, default initialized to (0+0i)
var variable complex128 // 128-bit complex number variable, default initialized to (0+0i)
var variabel error // error variable, default initialized to nil, used to represent an error condition



// ===============================
// ========== OPERATORS ==========
// ===============================

// Arithmetic Operators
var addtion = 12 + 5 // addition operator, adds two operands of same numeric type or concatenates two string types
var subtraction = 12 - 5 // subtraction operator, subtracts two operands of same numeric type
var multiplication = 12 * 5 // multiplication operator, multiplies two operands of same numeric type
var division = 12 / 5 // division operator, divides two operands of same numeric type, if both operands are integers the result will be an integer with the decimal part truncated
var modulus = 12 % 5 // modulus operator, returns remainder of division of two operands of same integer type
variable++ // increment operator, increases value of variable by 1

// Comparison Operators
var equal = 2 == 2 // equal operator, returns true if both operands are equal
var not_equal = 2 != 2 // not equal operator, returns true if both operands are not equal
var greater_than = 2 > 2 // greater than operator, returns true if the left operand is greater than the right operand
var less_than = 2 < 2 // less than operator, returns true if the left operand is less than the right operand
var greater_than_or_equal = 2 >= 2 // greater than or equal operator, returns true if the left operand is greater than or equal to the right operand
var less_than_or_equal = 2 <= 2 // less than or equal operator, returns true if the left operand is less than or equal to the right operand

// Logical Operators
var logical_and = true && false // logical AND operator, returns false if first operand is false or both operands are false, otherwise returns true
var logical_or = true || false // logical OR operator, returns true if first operand is true or both operands are true, otherwise returns false
var logical_not = !true // logical NOT operator, returns true if operand is false, otherwise returns false
variable-- // decrement operator, decreases value of variable by 1

// Bitwise Operators
var bitwise_and = 12 & 5 // bitwise AND operator, performs bitwise AND on two operands of same integer type
var bitwise_or = 12 | 5 // bitwise OR operator, performs bitwise OR on two operands of same integer type
var bitwise_xor = 12 ^ 5 // bitwise XOR operator, performs bitwise XOR on two operands of same integer type
var bit_clear = 12 &^ 5 // bit clear operator, performs bit clear on two operands of same integer type, clears the bits of the first operand which are set to 1 in the second operand
var left_shift = 12 << 2 // left shift operator, shifts the bits of the first operand to the left by the number of bits specified by the second operand, fills the vacated bits with 0
var right_shift = 12 >> 2 // right shift operator, shifts the bits of the first operand to the right by the number of bits specified by the second operand, fills the vacated bits with 0 for unsigned integers and with the sign bit for signed integers

// Assignment Operators
variable += 5 // addition assignment operator, adds the right operand to the left operand and assigns the result to the left operand
variable -= 5 // subtraction assignment operator, subtracts the right operand from the left operand and assigns the result to the left operand
variable *= 5 // multiplication assignment operator, multiplies the left operand by the right operand and assigns the result to the left operand
variable /= 5 // division assignment operator, divides the left operand by the right operand and assigns the result to the left operand
variable %= 5 // modulus assignment operator, takes the modulus of the left operand by the right operand and assigns the result to the left operand
variable &= 5 // bitwise AND assignment operator, performs bitwise AND on the left operand with the right operand and assigns the result to the left operand
variable |= 5 // bitwise OR assignment operator, performs bitwise OR on the left operand with the right operand and assigns the result to the left operand
variable ^= 5 // bitwise XOR assignment operator, performs bitwise XOR on the left operand with the right operand and assigns the result to the left operand
variable &^= 5 // bit clear assignment operator, performs bit clear on the left operand with the right operand and assigns the result to the left operand
variable <<= 5 // left shift assignment operator, performs left shift on the left operand by the number of bits specified by the right operand and assigns the result to the left operand
variable >>= 5 // right shift assignment operator, performs right shift on the left operand by the number of bits specified by the right operand and assigns the result to the left operand



// ========================================
// ========== CONTROL STATEMENTS ==========
// ========================================

// If-Else If-Else Statement
if variable > 19 { // if statement, { must be in the same line as the if keyword, will execute its block if the condition is true
	fmt.print("true condition") // code to execute if the condition is true
} else if variable < 10 { // else if statement, else if must be in the same line as the closing } of the previous block and { must be in the same line as the else if keyword, will check the condition if the previous if or else if conditions were false and execute its block if the condition is true
	fmt.print("else if condition") // code to execute if the else if condition is true
} else { // else statement, else must be in the same line as the closing } of the previous block and { must be in the same line as the else keyword, will execute its block if all previous if and else if conditions were false
	fmt.print("else condition") // code to execute if all previous conditions were false
} 

// Switch Statement with Expression
switch variable { // switch statement with an expression, { must be in the same line as the switch keyword, will compare the expression with the cases and execute the block of the first matching case
case 10: // case statement, case must be in the same line as the case keyword, will execute its block if the expression matches the case value
	fmt.print("case 10") // code to execute if the case matches
	fallthrough // fallthrough statement, used to transfer control to the next case, must be the last statement in the case block, will not check the next case condition
case 20, 30: // case statement with multiple values
	fmt.print("case 20 or 30") // code to execute if the case matches any of the values
default: // default case, default must be in the same line as the default keyword, will execute its block if no other case matches
	fmt.print("default case") // code to execute if no other case matches
}

// Switch Statement with Initialization
switch variable2 := 10; variable2 { // switch statement with initialization, same as above with initialization statement that must be in the same line as the switch keyword and separated from the expression by a semicolon
}

// Switch Statement without Expression
switch { // switch statement without an expression, equivalent to switch true, will execute the block of the first case whose condition is true, same as if statement
case variable > 19: // case statement with a condition, will execute its block if the condition is true
	fmt.print("case variable > 19")
	fallthrough // fallthrough statement, same as in switch with expression
case variable < 10: // case statement with another condition
	fmt.print("case variable < 10")
default: // default case, will execute its block if no other case condition is true
	fmt.print("default case")
}



// =====================================
// ========== LOOP STATEMENTS ==========
// =====================================

// Simple For Loop
for i := 0; i < 10; i++ { // for loop with initialization and condition and post statement separated by semicolons, { must be in the same line as the for keyword, will execute its block as long as the condition is true
	fmt.print(i) // code to execute in each iteration of the loop 
	if i == 5 { 
		break // break statement, used to exit from the loop
	} else if i == 3 {
		continue // continue statement, used to skip the current iteration and go to the next one
	}
}

// For Loop with just Condition
for variable < 10 { // for loop with just a condition, equivalent to while loop, will execute its block as long as the condition is true, { must be in the same line as the for keyword, continune and break work the same as in simple for loop
	fmt.print(variable) // code to execute in each iteration of the loop
}

// Forever Loop
for { // for with no condition, will execute its block forever, { must be in the same line as the for keyword, equivalent to for true, continune and break work the same as in simple for loop
	break
}

// For Range Loop

// Labeled For Loop



// ===============================
// ========== FUNCTIONS ==========
// ===============================

func function(param1 int, param2 string) float32 { // function declaration with parameters which can be 0, and a return type which can be omitted the function will return nothing, { must be in the same line as the function declaration
	return 0.0 // return statement, can be used to return a value and exit from the function, value can be omitted in void function to exit directly from it 
}
func function_with_multiple_return() (int, string) { // function declaration with multiple return types separated by a comma and enclosed in parentheses
	return 0, "" // return statement with multiple return values separated by a comma
}
var var1, var2 = function_with_multiple_return() // initialization of multiple variables with the return values of a function